File: procrustes.cpp

void rotation(){
    cv::Mat rot = cv::getRotationMatrix2D(cv::Point2f(0,0), 60, 1);

    std::cout << rot << std::endl;
    std::cout << rot.dims << " " << rot.cols << " " << rot.rows << std::endl;
}


void procrustes(cv::Mat X, cv::Mat Y){
    
    // X, Y
    // [2 x 3] 変換する3点の行列, rows = 3, cols = 2
    // 
    int n = X.rows;
    int m = X.cols;
    int ny = Y.rows;
    int my = Y.cols;
   
    // 列平均（それぞれの x, y 座標の平均値を算出）
    // [2 x 1]
    //
    cv::Mat muX_, muY_;
    cv::reduce(X, muX_, 0, cv::REDUCE_AVG, CV_32F);
    cv::reduce(Y, muY_, 0, cv::REDUCE_AVG, CV_32F);
    
    // [2 x 3] 
    cv::Mat muX, muY;
    for(int idx=0; idx<3; idx++){
        muX.push_back(muX_);
        muY.push_back(muY_);
    }
    std::cout << "mux = " << muX << std::endl;

    cv::Mat X0, Y0;
    X0 = X - muX;
    Y0 = X - muY;

    std::cout << "X0 = " << X0 << std::endl;

    // 要素毎の積
    cv::Mat X0_2 = X0.mul(X0); 
    cv::Mat Y0_2 = Y0.mul(Y0); 

    double ssX = cv::sum(X0_2)[0];
    double ssY = cv::sum(Y0_2)[0];

    double normX = std::sqrt(ssX);
    double normY = std::sqrt(ssY);

    std::cout << "normX = " << normX << std::endl;

    X0 /= normX;
    Y0 /= normY;

    


}

File: procrustes.py

import numpy as np

def procrustes(X, Y, scaling=True, reflection='best'):

    n,m = X.shape
    ny,my = Y.shape

    muX = X.mean(0)
    muY = Y.mean(0)

    X0 = X - muX
    Y0 = Y - muY

    ssX = (X0**2.).sum()
    ssY = (Y0**2.).sum()

    # centred Frobenius norm
    normX = np.sqrt(ssX)
    normY = np.sqrt(ssY)

    # scale to equal (unit) norm
    X0 /= normX
    Y0 /= normY

    if my < m:
        Y0 = np.concatenate((Y0, np.zeros(n, m-my)),0)

    # optimum rotation matrix of Y
    A = np.dot(X0.T, Y0)
    U,s,Vt = np.linalg.svd(A,full_matrices=False)
    V = Vt.T
    T = np.dot(V, U.T)

    traceTA = s.sum()


	# ------------------------
	# scaling
	# ------------------------
	
    # optimum scaling of Y
    b = traceTA * normX / normY

    # standarised distance between X and b*Y*T + c
    d = 1 - traceTA**2

    # transformed coords
    Z = normX*traceTA*np.dot(Y0, T) + muX


    # transformation matrix
    if my < m:
        T = T[:my,:]
    c = muX - b*np.dot(muY, T)
   
 
    #transformation values 
    tform = {'rotation':T, 'scale':b, 'translation':c}
   
    return d, Z, tform



if __name__ == "__main__" : 

    X = np.array([
        [],
        [],
        [],
    ])
    Y = np.random.randn(3)

